### PR TITLE
install: Keep relay available while upgrading

### PR DESCRIPTION
**Pull request by ktosiek**
This continues on the ideas from #607. By "downtime" here I mean "not accepting events" - web, smtp and background processes are out of scope.

This PR adds a `--minimize-downtime` option to install.sh. This options changes the behaviour of the script by:
1. keeping nginx and relay running until the very end,
2. disabling cleanup on exit and failure,
3. explicitly reloading nginx configuration,
4. and starting the whole cluster at the end.

This is a work in progress, and only tested on a local, non-production instance. But the results are promising: no downtime if relay version doesn't change, and only a second when it does. So far this was only tested with a curl loop, so I'm still not sure if Relay flushes the events to Sentry before getting recreated by `$dc up`.

### CODE REVIEW

**Comment by BYK** [comment]
@ktosiek - okay, I'm excited about this. What keeps this PR from being ready to review?

**Comment by ktosiek** [comment]
Testing - I still need to see how it will perform when relay is getting upgraded too.

With the current implementation web, Kafka and Redis would go down, and after a minute everything (including Relay) would get restarted. If Relay has no on-disk queue then this PR will only make things worse (clients would get a 200, but events would be dropped).

**Comment by BYK** [comment]
I am fairly sure Relay has at least an in-memory queue. I remember having heard some storage in Redis but we need to wait for @jan-auer for confirmation. He'll be back next week.

**Comment by jan-auer** [comment]
> If Relay has no on-disk queue then this PR will only make things worse

I'm afraid this is correct. If either Redis or Kafka are down, Relay will drop events. Likewise, while Relay is starting up and not authenticated, it will also drop events. Both of these are known issues. For the startup behavior, we are planning to implement a fix in the short term.

For Kafka or Redis being down we do not have fixes on the immediate roadmap. This is definitely on our radar, though, as we're looking into the queueing behavior at the moment.

**Comment by BYK** [comment]
Well, then we can at least restart these sequentially at the end? Like:

1. Restart all things Sentry
2. Restart Redis
3. Restart all things Kafka
4. Restart Snuba
5. Restart Relay
6. Restart Nginx

**Comment by ktosiek** [comment]
I think this is ready now - I've tested actually sending events, and every event that was accepted with 200 was visible in Sentry after upgrade (well, at least the count of events matched). Events were not being ingested for ~10 seconds - seems great for a single server setup :-)

The script I've used to send events is not included in the PR: https://gist.github.com/ktosiek/f35a51cd71b151aaf6ff7f1e981ea01f

File: install.sh
Comment (kohenkatz) [inline install.sh:337]: Can there be a better way to do this for those of us not exposing port 9000?

I'm using Traefik as an edge router, so I don't expose any ports from Sentry at all.

My suggestion would be to keep `http://localhost:9000` as the default but allow overriding it, either with an environment variable (preferred) or an argument to the script. If using an argument, maybe it could be `--minimize-downtime` for no change to the URL and `--minimize-downtime=https://sentry.example.com` to override it.

Comment (kohenkatz) [reply install.sh:337]: If you want, I can wait until this is merged and then open a new PR to add the behavior I need.

Comment (kohenkatz) [reply install.sh:337]: Related: #512

Comment (ktosiek) [reply install.sh:337]: Good catch! I don't really care for the proxy here, so it should probably run in a container and wait for `web:9000`.

Comment (BYK) [reply install.sh]: I think we can merge #512 before to make things easier for this PR, that said we can also do something like https://unix.stackexchange.com/a/45943/108960 replacing the `tail -f` part with `$dc logs -f web` and look for some uWSGI/Django success message and avoid curling and ports altogether.

File: install.sh
Comment (BYK) [inline install.sh:333]:
```suggestion
  $dc up -d --remove-orphans $($dc config --services | grep -v -e '^(nginx|relay)$')
```

File: install.sh
Comment (ktosiek) [reply install.sh]: I've changed it to waiting for `http://web:9000/_health/` from an alpine container - this way I don't have to think about the port, as I'm running inside the compose project's network. Plus this sheds the `curl` dependency on host :-)

File: install.sh
Comment (ktosiek) [reply install.sh]: Fixed in both places.

**Comment by BYK** [comment]
@ktosiek - what do you say, are we ready to go?

File: install.sh
Comment (BYK) [inline install.sh:31]:
```suggestion
        -h | --help) show_help; exit;;
        --minimize-downtime) MINIMIZE_DOWNTIME=1;;
        --) ;;
        *) echo "Unexpected argument: $1. Use --help for usage information."; exit 1;;
```

Comment (BYK) [reply install.sh:31]: Let's just use spaces for consistency :)

Comment (ktosiek) [reply install.sh]: I see the code mixes 2- and 4-space indents - do you think an .editorconfig (in a separate PR, of course) would help?

Comment (BYK) [reply install.sh]: >I see the code mixes 2- and 4-space indents - do you think an .editorconfig (in a separate PR, of course) would help?

Possibly, for vim users I think 😀 I think ultimately we'd need a formatter and a lint rule for this file.

>And if so, what's the preferred style?

I'd go with 2-spaces to be consistent with the `docker-compose.yml` file (although my Python brain wants 4-spaces)

File: install.sh
Comment (ktosiek) [reply install.sh:334]: This line is here only for one case: when the nginx configuration changes, but the container configuration doesn't. In that case docker-compose won't restart nginx at the final `$dc up -d`.

**Comment by ktosiek** [comment]
The test I've added only checks if a no-op upgrade works, but I don't have a good idea for how to do it differently.
I think this is ready to go, that's how I did my last production upgrade already ^_^

File: install.sh
Comment (kohenkatz) [reply install.sh]: @BYK `.editorconfig` is supported in every editor I know of, including VSCode, all IntelliJ-based editors, Atom, Eclipse, Notepad++, Netbeans, SublimeText, and more. You can also set different formats for different file types (or even different file names).

There are also already linters that check all files against the editorconfig configuration, including:

- https://github.com/editorconfig-checker/editorconfig-checker
- https://github.com/jedmao/eclint
- https://gitlab.com/greut/eclint

**Review by BYK — APPROVED** [review]

**Pull request merged**
